src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCodeLensProvider.ts

```diff
@@ -0,0 +1,130 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { isRazorCSharpFile, getRazorDocumentUri } from './RazorConventions';
+import { RazorLanguageFeatureBase } from './RazorLanguageFeatureBase';
+import { LanguageKind } from './RPC/LanguageKind';
+import { CancellationToken } from 'vscode-jsonrpc';
+
+class RazorCodeLens extends vscode.CodeLens {
+    constructor(
+        range: vscode.Range,
+        public uri: vscode.Uri,
+        public document: vscode.TextDocument,
+        command?: vscode.Command) {
+
+        super(range, command);
+    }
+}
+
+export class RazorCodeLensProvider
+    extends RazorLanguageFeatureBase
+    implements vscode.CodeLensProvider {
+
+    public async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken) {
+        try {
+            const razorDocument = await this.documentManager.getDocument(document.uri);
+            if (!razorDocument) {
+                return;
+            }
+
+            const csharpDocument = razorDocument.csharpDocument;
+
+            // Get all the code lenses that applies to our projected C# document.
+            const codeLenses = await vscode.commands.executeCommand<vscode.CodeLens[]>(
+                'vscode.executeCodeLensProvider',
+                csharpDocument.uri) as vscode.CodeLens[];
+
+            // Re-map the CodeLens locations to the original Razor document.
+            const remappedCodeLenses = new Array<vscode.CodeLens>();
+            for (const codeLens of codeLenses) {
+                const result = await this.serviceClient.mapToDocumentRange(
```

Comment:
Food for thought: I've noticed a general trend of us calling map to document range in for loops. We may want to keep in the back of our minds the possibility of creating an optimized variant which takes multiple inputs.

Reply:
Yeah that might be a good optimization to have. This also means this API should accept non-razor documents as well and should gracefully no-op.

Reply:
Is it possible for one of these documents to refer to a non-razor file? My understanding is that things defined within a Razor file can be accessible from regular C#

Reply:
All of the items in the `codeLenses` array will point to the same file (the background C# document for the current Razor document). CodeLenses are locations within **this** file where the gray links should be placed. The one you are referring to is what shows up after someone clicks on a code lens which can point to any file but that is handled in `resolveCodeLens`.